File: sentimental-credit/credit.py
from cs50 import get_string


def main():
    card = get_string("Number: ")
    n = int(card)
    if check_sum(card):
        if len(card) == 15 and (
            card.startswith("34") == True or card.startswith("37") == True
        ):
            print("AMEX\n")
        elif (len(card) == 13 or len(card) == 16) and card.startswith("4") == True:
            print("VISA\n")
        elif len(card) == 16 and (n > (50 * 10**14) and n < (56 * 10**14)):
            print("MASTERCARD\n")
        else:
            print("INVALID\n")
    else:
        print("INVALID\n")


def check_sum(card):
    double_sum = 0
    pure_sum = sum(int(i) for i in card[len(card) - 1 :: -2])
    for i in card[len(card) - 2 :: -2]:
        temp_sum = 2 * int(i)
        temp_string = str(temp_sum)
        # print(i, temp_sum, temp_string)
        if len(temp_string) == 2:
            temp_sum = int(temp_string[0]) + int(temp_string[1])
        double_sum = double_sum + temp_sum
    checksum = pure_sum + double_sum
    # print(pure_sum, double_sum, checksum)
    if checksum % 10 == 0:
        return 1
    else:
        return 0


main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Prompt for input
    long card;
    long n;
    string type;
    bool valid = false;
    while (valid == false)
    {
        int count = 0;
        card = get_long("Enter valid card number: ");
        n = card;
        long n2 = n;
        long n1 = n;
        while (n != 0)
        {
            n2 = n1;
            n1 = n;
            n = n / 10;
            count++;
        }
        if ((count == 15) && ((n2 == 34) || (n2 == 37)))
        {
            type = "AMEX";
            valid = true;
        }
        else
        {
            if (((count == 16) || (count == 13)) && (n1 == 4))
            {
                type = "VISA";
                valid = true;
            }
            else
            {
                if ((count == 16) && ((n2 > 50) && (n2 < 56)))
                {
                    type = "MASTERCARD";
                    valid = true;
                }
                else
                {
                    printf("INVALID\n");
                    return (0);
                }
            }
        }
    }
    // Calculate Checksum
    int checksum = 0;
    long remainder = 0;
    long remainder1 = 0;
    long remainder2 = 0;
    int count = 0;
    n = card;
    while (n != 0)
    {
        remainder = n % 10;
        count = count + 1;
        if (count % 2 == 0)
        {
            remainder = remainder * 2;
            if (remainder < 10)
            {
                checksum = checksum + remainder;
            }
            else
            {
                remainder1 = remainder / 10;
                remainder2 = remainder % 10;
                checksum = checksum + remainder1 + remainder2;
            }
        }
        else
        {
            checksum = checksum + remainder;
        }
        n = n / 10;
    }
    if (checksum % 10 == 0)
    {
        printf("%s\n", type);
        return (0);
    }
    else
    {
        printf("INVALID\n");
        return (0);
    }
}
